File: sdks/nodejs/src/payments.ts
import { CloudPoolClient } from './index';

export interface ConnectGatewayConfig {
  provider: string;
  apiKey: string;
  webhookSecret?: string;
}

export interface CreateCheckoutConfig {
  planId: string;
  successUrl: string;
  cancelUrl: string;
  metadata?: Record<string, string>;
}

export class PaymentsClient {
  constructor(private readonly client: CloudPoolClient) {}

  public async listGateways<T = unknown>(): Promise<T[]> {
    const response = await this.client.request(
      'GET',
      '/api/v1/payments/gateways'
    );
    return response.json();
  }

  public async connectGateway<T = unknown>(config: ConnectGatewayConfig): Promise<T> {
    const response = await this.client.request(
      'POST',
      '/api/v1/payments/gateways',
      {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(config),
      }
    );
    return response.json();
  }

  public async disconnectGateway<T = unknown>(id: string): Promise<void> {
    await this.client.request(
      'DELETE',
      `/api/v1/payments/gateways/${encodeURIComponent(id)}`
    );
  }

  public async listPlans<T = unknown>(): Promise<T[]> {
    const response = await this.client.request(
      'GET',
      '/api/v1/payments/plans'
    );
    return response.json();
  }

  public async listInvoices<T = unknown>(params?: {
    limit?: number;
    status?: string;
  }): Promise<T[]> {
    const searchParams = new URLSearchParams();
    if (params?.limit) searchParams.set('limit', String(params.limit));
    if (params?.status) searchParams.set('status', params.status);
    const query = searchParams.toString();
    const response = await this.client.request(
      'GET',
      `/api/v1/payments/invoices${query ? '?' + query : ''}`
    );
    return response.json();
  }

  public async createCheckout<T = unknown>(config: CreateCheckoutConfig): Promise<T> {
    const response = await this.client.request(
      'POST',
      '/api/v1/payments/checkout',
      {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(config),
      }
    );
    return response.json();
  }

  public async listUsage<T = unknown>(params?: {
    startDate?: string;
    endDate?: string;
  }): Promise<T[]> {
    const searchParams = new URLSearchParams();
    if (params?.startDate)
      searchParams.set('startDate', params.startDate);
    if (params?.endDate) searchParams.set('endDate', params.endDate);
    const query = searchParams.toString();
    const response = await this.client.request(
      'GET',
      `/api/v1/payments/usage${query ? '?' + query : ''}`
    );
    return response.json();
  }
}

File: sdks/rust/src/payments.rs
use reqwest::Method;
use serde_json::Value;

use crate::client::CloudPoolClient;
use crate::errors::CloudPoolError;

/// Client for payment management: gateways, plans, invoices, checkout, and usage.
pub struct PaymentsClient<'a> {
    client: &'a CloudPoolClient,
}

impl<'a> PaymentsClient<'a> {
    /// Create a new payments client.
    pub fn new(client: &'a CloudPoolClient) -> Self {
        Self { client }
    }

    /// List all connected payment gateways.
    pub async fn list_gateways(&self) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::GET, "payments/gateways"))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// Connect a new payment gateway.
    pub async fn connect_gateway(&self, config: Value) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(
                self.client
                    .request(Method::POST, "payments/gateways")
                    .json(&config),
            )
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// Disconnect (remove) a payment gateway by ID.
    pub async fn disconnect_gateway(&self, gateway_id: &str) -> Result<Value, CloudPoolError> {
        let path = format!("payments/gateways/{}", gateway_id);
        let res = self
            .client
            .send_with_retry(self.client.request(Method::DELETE, &path))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// List all available subscription plans.
    pub async fn list_plans(&self) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::GET, "payments/plans"))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// List all invoices for the authenticated account.
    pub async fn list_invoices(&self) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::GET, "payments/invoices"))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// Create a checkout session for a given plan.
    pub async fn create_checkout(&self, plan_id: &str) -> Result<Value, CloudPoolError> {
        let body = serde_json::json!({ "planId": plan_id });
        let res = self
            .client
            .send_with_retry(
                self.client
                    .request(Method::POST, "payments/checkout")
                    .json(&body),
            )
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// List usage records for the current billing period.
    pub async fn list_usage(&self) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::GET, "payments/usage"))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }
}
